### PR TITLE
Make embeddings from data constant per store-item group.

### PR DESCRIPTION
**Pull request by nnistelrooij**
All the data constant per store-item group are now one-hot vectors, where each type of data (week, weekday, month day, month, year, and event type) are normalized to z-scores.
The prices and SNAP data are also normalized to z-scores, but the sales not, as they are used as targets.

I added a linear layer with ReLU non-linearity to make embeddings from the concatenated one-hot vectors. This layer is run once per iteration in the forward pass of `Model`, so it probably doesn't hurt efficiency much. The dimensionality of 20 might be a bit overkill for the 6 data types, but it doesn't really matter, because it's efficient.

### CODE REVIEW

**Comment by nnistelrooij** [comment]
Whelp, with the embeddings it's actually faster. Nice!

File: utils/data.py [hunk 1]
```diff
@@ -124,12 +145,13 @@ def _sell_prices(calendar, prices):
         prices = prices.fillna(0)
         prices = prices[calendar['wm_yr_wk']]
 
-        # return prices in (num_days, num_groups) shape
+        # return normalized prices in (num_days, num_groups) shape
+        prices = (prices - 3.507094282300549) / 3.5222671176612437
```
Comment (MauiGonzo) [inline utils/data.py:149]: I might have been asked before, but where do these constants come from?

**Review by MauiGonzo — APPROVED** [review]

File: nn.py [hunk 2]
```diff
@@ -473,16 +469,16 @@ def reset_hidden(self):
...
+    def forward(self, day, t_day, items, t_items):
         """Forward pass of the submodel.
 
         Args:
             day   = [torch.Tensor] inputs constant per store-item group
                 The shape should be (1, seq_len, num_const).
-            items = [torch.Tensor] inputs different per store-item group
-                The shape should be (1, seq_len, num_groups, num_var).
             t_day   = [torch.Tensor] targets different per store-item group
                 The shape should be (1, horizon, num_const).
```
Comment (rusane) [inline nn.py]: Volgens mij moet dit `targets constant per store-item group` zijn

File: nn.py [hunk 3]
```diff
@@ -522,16 +518,21 @@ def __init__(self, num_models, dropout, device):
...
         """
-        global num_groups
+        global num_const
 
         super(Model, self).__init__()
 
-        min_model_groups = num_groups // num_models
-        num_extra_groups = num_groups % num_models
+        # make linear layer for embeddings constant per store-item group
+        self.fc = nn.Linear(ForecastDataset.num_const, num_const).to(device)
```
Comment (rusane) [inline nn.py:526]: Is `ForecastDataset.num_const` anders dan `num_const`? Ik dacht je de `global num_const` in de `ForecastDataset` hebt gestopt en dat je dan niet meer `global num_const` nodig hebt (i.e., ze zijn hetzelfde).

File: nn.py [hunk 4]
```diff
@@ -555,24 +556,29 @@ def forward(self, day, items, t_day, t_items):
         Args:
             day   = [torch.Tensor] inputs constant per store-item group
                 The shape should be (1, seq_len, num_const).
-            items = [torch.Tensor] inputs different per store-item group
-                The shape should be (1, seq_len, num_groups, num_var).
             t_day   = [torch.Tensor] targets different per store-item group
                 The shape should be (1, horizon, num_const).
```
Comment (rusane) [inline nn.py]: Zelfde als eerder: `targets constant per store-item group`

File: utils/data.py [hunk 5]
```diff
@@ -33,57 +40,68 @@ def __init__(self, calendar, prices, sales, seq_len=1, horizon=1):
...
 
+    @staticmethod
+    def _normalize(input):
+        """Normalizes one-hot vectors to z-scores."""
+        num_classes = input.shape[1]
+
+        mean = 1 / num_classes
+        std = math.sqrt((num_classes - 1) * mean**2)
+
+        return (input - mean) / std
```
Comment (rusane) [inline utils/data.py:73]: Ik vind het nog steeds een beetje vreemd om one hot vectors the normaliseren omdat het eigenlijk gewoon meerdere binary variables zijn. Ik heb even kort gezocht op internet en het lijkt niet gebruikelijk, heb je zelf wel eerder gezien dat one hot vectors worden genormaliseerd?

File: utils/data.py (hunk 1 above)
Comment (nnistelrooij) [reply utils/data.py:149]: I took the mean and standard deviation of the prices of all 1969 days.

File: nn.py (hunk 2 above)
Comment (nnistelrooij) [reply nn.py]: Yup

File: nn.py (hunk 3 above)
Comment (nnistelrooij) [reply nn.py:526]: Alle one-hot vectors bij elkaar zijn 114 elementen (`ForecastDataset.num_const`) en na de FC laag hebben we embeddings van 20 elementen (`nn.num_const`).

File: utils/data.py (hunk 5 above)
Comment (nnistelrooij) [reply utils/data.py:73]: There isn't really a consensus, basically: "do what works best". The next run will be without normalizing to see what works better

**Review by rusane — APPROVED** [review]

**Comment by nnistelrooij** [comment]
Scrap this whole idea. Without embeddings it's worse on the train set, but better on the validation set.
